enhancements/machine-api/out-of-tree-provider-support.md

```diff
@@ -0,0 +1,232 @@
+
+`Cloud-controller` will be provisioned in the `openshift-cloud-controller-manager` namespace.
+
+Preparation steps, common for all providers:
+
+- Add support for recognition of the `external` option in [library-go](https://github.com/openshift/library-go/blob/master/pkg/operator/configobserver/cloudprovider/observe_cloudprovider.go) which will disable config parsing and add support for `external` field in the `Infrastructure` resource.
+- Port the templates of `cloud-controller` from upstream repositories into proposed https://github.com/openshift/cluster-cloud-controller-manager-operator repository. Those templates will belong to the newly created `openshift-cloud-controller-manager` namespace and be managed with an operator.
+- Operator will be deployed in the `openshift-cloud-controller-manager-operator` namespace.
+
+Each provider implementation will need to do the following minimal set of actions in order to make out-of-tree implemetation work:
+
+- Use `external` flag in the `kube-controller-manager` pod instead of the previously used provider. Remove the `cloud-config` option from the template.
+- Use `external` flag in the `kubelet` pod. Remove the `cloud-config` option from the template.
+- Provision a `cloud-controller-manager` `DaemonSet` in the `openshift-cloud-controller-manager` namespace.
+
+#### API resource changes
+
+In order to achieve desired cluster configuration with `external` cloud components, we need to add a new field into `Infrastructure` [resource](https://github.com/openshift/api/blob/c2f7aea5d89ee87d7510b50bdb08dde9028a53ef/config/v1/types_infrastructure.go#L11). With the addition, a `status.platformStatus.external` field will determine the architectural change in the cluster configuration. Other components will be able to consume the configuration same way as before, only with a minor change of the resource processing.
+
+```go
+type PlatformStatus struct {
+  ...
+	// External is a value used in infrastucture automaton to determine component-specific
+	// behavior for the PlatformType value. Components depending on this variable will
+	// ignore cloud specific detailes associated with the given platform type,
+	// and will concider the value of the infrastructure to be "external" - essentially
+	// disabling any further evaluation.
+	//
+	// This will allow to delegate task of processing platform specific bits
+	// to another component.
+	// +optional
+  External bool `json:"external"`
+  ...
+```
+
+To ensure a smooth transition from the `in-tree` selection, `library-go` [implementation](https://github.com/openshift/library-go/blob/master/pkg/operator/configobserver/cloudprovider/observe_cloudprovider.go) will include an additional argument for the `cloudProviderObserver` structure, which will allow to preserve initial configuration strategy, until the new `observeExternal` flag will be set to `false`:
```

Comment:
who will set observeExternal and when?

Reply:
There are a couple of PRs I opened to test the approach. This option will be set on `false` for all components currently watching the `cloud-config` file and relying on `library-go` implementation once they will be merged:
- https://github.com/openshift/library-go/pull/895
- https://github.com/openshift/cluster-kube-apiserver-operator/pull/953
- https://github.com/openshift/cluster-kube-controller-manager-operator/pull/450

These I assume should be merged into 4.7 branch to introduce the transitional phase.

Reply:
I think this paragraph needs an update now we are leaning towards this temporary annotation approach

Reply:
It still applies, as it would be good to reuse the logic from the library-go implementation in both `kcm` and `ccm` operators. It will only rely on the annotation in this case.